modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointMarkersStorage.java

```diff
@@ -535,4 +608,100 @@ public static String checkpointFileName(CheckpointEntry cp, CheckpointEntryType
     public CheckpointHistory history() {
         return cpHistory;
     }
+
+    /**
+     * See {@link CheckpointHistory#removeFromEarliestCheckpoints}.
+     *
+     * @param grpId Group id.
+     * @return Checkpoint entry.
+     */
+    public CheckpointEntry removeFromEarliestCheckpoints(Integer grpId) {
+        CheckpointEntry entry = cpHistory.removeFromEarliestCheckpoints(grpId);
+
+        onEarliestCheckpointMapChanged();
+
+        return entry;
+    }
+
+    /**
+     * Handles changes in the earliest checkpoint map in the {@link CheckpointHistory}. Creates a snapshot of that
+     * map if threshold for changes has been reached.
+     */
+    void onEarliestCheckpointMapChanged() {
+        boolean createCheckpoint =
+            checkpointSnapshotCounter.getAndUpdate(old -> (old + 1) % EARLIEST_CP_CHANGES_THRESHOLD) == 0;
+
+        if (createCheckpoint) {
+            Runnable runnable = () -> {
+                EarliestCheckpointMapSnapshot snapshot;
+
+                lock.readLock();
```

Comment:
It is a suspicious lock.

Reply:
Well I have to lock on the checkpoint lock so I can make a snapshot of checkpoint map